pkg/virt-launcher/virtwrap/device/sriov/hostdev_test.go

```diff
@@ -0,0 +1,252 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2020 Red Hat, Inc.
+ *
+ */
+
+package sriov_test
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	v1 "kubevirt.io/client-go/api/v1"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/device/sriov"
+)
+
+var _ = Describe("SRIOV HostDevice", func() {
+	It("creates no device given no interfaces", func() {
+		vmi := &v1.VirtualMachineInstance{}
+
+		Expect(sriov.CreateHostDevices(vmi)).To(BeEmpty())
+	})
+
+	It("creates no device given no SRIOV interfaces", func() {
+		iface := v1.Interface{}
+		iface.Masquerade = &v1.InterfaceMasquerade{}
+		vmi := &v1.VirtualMachineInstance{}
+		vmi.Spec.Domain.Devices.Interfaces = []v1.Interface{iface}
+
+		Expect(sriov.CreateHostDevices(vmi)).To(BeEmpty())
+	})
+
+	It("fails to create device given no available host PCI", func() {
+		iface := newSRIOVInterface("test")
+		vmi := &v1.VirtualMachineInstance{}
+		vmi.Spec.Domain.Devices.Interfaces = []v1.Interface{iface}
+		_, err := sriov.CreateHostDevices(vmi)
+
+		Expect(err).To(HaveOccurred())
+	})
+
+	It("fails to create a device given bad host PCI address", func() {
+		net := newNetworkData("net1", newResourceData("resource1", "0bad0pci0address0"))
+
+		iface := newSRIOVInterface(net.Name)
+		vmi := &v1.VirtualMachineInstance{}
+		vmi.Spec.Domain.Devices.Interfaces = []v1.Interface{iface}
+
+		var err error
+		env := []envData{net.ResourceEnv, net.DeviceEnv}
+		withEnvironmentContext(env, func() {
```

Comment:
I have a solution that does not require an interface, but it still has the API messier.
The API from the `manager` will look like this:
```golang
sriovInterfaces := sriov.FilterVMISRIOVInterfaces(vmi.Spec.Domain.Devices.Interfaces)
sriovDevices, err := sriov.CreateHostDevices(sriovInterfaces, sriov.NewPCIAddressPool(sriovInterfaces))
```

Instead of the current:
```golang
sriovDevices, err := sriov.CreateHostDevices(vmi)
```

Do you prefer this?

Reply:
Thank you for insisting, I think it is done now.